### DISCUSSION TITLE
Display 3D Voxelized Topology Optimization Results

### DISCUSSION THREAD

**Discussion by bhttchr6** [marcomusy/vedo#546] [Q&A] [answered]
Hi,
I started using vedo recently and have already got hooked to its features and ease of use. I am currently exploring its use for plotting topology optimization results in 3D. 
I have a density could (scalar numpy array) that represent the density values at the element centre. My goal is to use the density value to color all the surfaces of the lego blocks that I want to use to represent each voxel/mesh cell. Currently I am unable to do that using vedo as each surface is getting colored with a different color also some vovels are not represented at all.
I believe it should be pretty straightforward but maybe due to lack of experience with vedo, I am unable to represent it.
Any guidance and help, along with any example, regarding this will be very helpful.
Thanks,
Anurag

**Comment by marcomusy** [comment]
Hi Anurag, I'm not sure I understand what you are trying to do... you want to visualize a custom/predefined subset of voxels as small cubes? and color it based on the voxel value?

**Reply by bhttchr6** [reply]
Hi,
Sorry for confusing you. 
What I mean is that, say,  I have a set of 3D finite elements (e.g number of elements in x-dir = 3 (say), in y-dir = 2 and z-dir = 2) and each of these elements have a scalar density value associate with them which is represented as a numpy array. I want that each of these elements be represented as a lego block with all the surfaces of the lego block having the same color represented by the scalar elemental density value. Also, the elements which have a density value of say 0, be not plotted at all.
Hope I have been able to make myself a bit clearer.
Thanks for such a prompt response.

Regards,
Anurag

**Comment by marcomusy** [comment, marked as answer]
Uhm, maybe you wish to use a `Glyph` object to represent the voxels, e.g.:

```python
import numpy as np
from vedo import *

X, Y, Z = np.mgrid[:30, :30, :30]
scalar_field = ((X-0)**2 + (Y-15)**2 + (Z-15)**2)/225

vol = Volume(scalar_field)

pcl = vol.topoints().ps(4)
# pcl.print()
# pcl.pointdata.select('input_scalars')
# pcl.pointdata['myscals'] = pcl.pointdata['input_scalars']+10

marker = Cube().scale(0.2)
glyphed_pcl = Glyph(pcl, marker, scaleByScalar=True).cmap('jet')
glyphed_pcl.addScalarBar()

show(pcl, glyphed_pcl, N=2, axes=1)
```

![Screenshot from 2021-11-29 23-41-04](https://user-images.githubusercontent.com/32848391/143954157-ea3ef3b5-f864-4951-b60b-8b0b5b5710fe.png)

**Comment by bhttchr6** [comment]
This works like a charm!. Thanks a lot!. Just one minor thing that I would ask: When I do 'scaleByScalar = False', it plots something like this.
![image](https://user-images.githubusercontent.com/47575638/143956888-78ba1474-936a-4d2e-9274-8cfd068480c0.png)


and when I do 'scaleByScalar = True', it plots:
![image](https://user-images.githubusercontent.com/47575638/143957067-5478116c-26e3-453a-a1d4-86db71595c5a.png)
However I want the blue part to be kept, as that is my solution while the red-part to be omitted. Is there a command to show one color and not the other.
Thanks again!
Anurag

**Reply by marcomusy** [reply]
Hi Anurag
`scaleByScalar` does what the name says: scales the cube marker by the value of the scalar, that is why the blue (zeros) cubes disappear.
You can manually create a new array which is inverted and scale with that (follow the commented out lines in the above example), you can also invert the colormap (eg `.cmap('jet_r')`).

**Reply by bhttchr6** [reply]
Thanks a lot!. Made it work.
